### PR TITLE
[SPARK-34357][SQL] Map JDBC SQL TIME type to TimestampType with time portion fixed regardless of timezone

### PR DESCRIPTION
**Pull request by saikocat**
### What changes were proposed in this pull request?

Due to user-experience (confusing to Spark users - java.sql.Time using milliseconds vs Spark using microseconds; and user losing useful functions like hour(), minute(), etc on the column), we have decided to revert back to use TimestampType but this time we will enforce the hour to be consistently across system timezone (via offset manipulation) and date part fixed to zero epoch.

Full Discussion with Wenchen Fan Wenchen Fan regarding this ticket is here https://github.com/apache/spark/pull/30902#discussion_r569186823

### Why are the changes needed?

Revert and improvement to sql.Time handling

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests and integration tests

### CODE REVIEW

**Comment by saikocat** [comment]
CC: @cloud-fan @sarutak @skestle - regarding the updated changes to respect vendor db integrations

**Comment by sarutak** [comment]
ok to test.

**Comment by sarutak** [comment]
@saikocat Have you confirmed all the integration tests like `PostgresIntegrationSuite` passes?
Jenkins and GA don't run them so we need to confirm by ourselves.

**Comment by saikocat** [comment]
> @saikocat Have you confirmed all the integration tests like `PostgresIntegrationSuite` pass?
> Jenkins and GA don't run them so we need to confirm by ourselves.

Yup, I learned from my previous mistake. Though for mssql, I couldn't run it at all due to images not available/access denied (both GA and latest) :< the rest IT all passed.

**Comment by sarutak** [comment]
> Yup, I learned from my previous mistake. All IT all passed.
>
> Edit: manual loaded the docker image for mssql works

Thanks. Anyway, I'll run all the integration tests before this change merged.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 1]
```diff
@@ -470,6 +455,27 @@ object JdbcUtils extends Logging {
...
+    // time portion is time of day, with no reference to a particular calendar,
+    // time zone or date, with a precision till microseconds.
+    // It stores the number of milliseconds after midnight, 00:00:00.000000
+    case TimestampType if metadata.contains("logical_time_type") =>
+      (rs: ResultSet, row: InternalRow, pos: Int) => {
+        val rawTime = rs.getTime(pos + 1)
+        if (rawTime != null) {
+          val localTimeMicro = TimeUnit.NANOSECONDS.toMicros(rawTime.toLocalTime().toNanoOfDay())
+          val localTimeMillis = DateTimeUtils.microsToMillis(localTimeMicro)
+          val timeZoneOffset = TimeZone.getDefault match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:469]: In most cases, session timezone should be the same as JVM default timezone, but ideally we should use session timezone for datetime operations: `SQLConf.get.sessionLocalTimeZone`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 2]
```diff
@@ -470,6 +455,27 @@ object JdbcUtils extends Logging {
...
+            case zoneInfo: ZoneInfo => zoneInfo.getOffsetsByWall(localTimeMillis, null)
+            case timeZone: TimeZone => timeZone.getOffset(localTimeMillis - timeZone.getRawOffset)
+          }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:472]: Seems we can just do `DateTimeUtils.toUTCTime(localTimeMicro, SQLConf.get.sessionLocalTimeZone)`

Comment (saikocat) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:472]: Superb! This is really more elegant. Thanks for this suggestion!

**Review by cloud-fan — APPROVED** [review]

**Comment by sarutak** [comment]
Confirmed all the JDBC integration tests pass.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**